File: 042/trapping-rain-water.go
package main

import "fmt"

func trap(height []int) int {
	size := len(height)
	if size < 2 {
		return 0
	}

	leftMaxHeight, rightMaxHeight := make([]int, size), make([]int, size)
	leftMaxHeight[0], rightMaxHeight[size-1] = height[0], height[size-1]

	for i := 1; i < size; i++ {
		leftMaxHeight[i] = max(leftMaxHeight[i-1], height[i])
		rightMaxHeight[size-1-i] = max(rightMaxHeight[size-i], height[size-1-i])
	}

	water := 0
	for i := 0; i < size; i++ {
		water += min(leftMaxHeight[i], rightMaxHeight[i]) - height[i]
	}

	return water
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	height := []int{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}
	fmt.Println(trap(height))
}

File: 042/TrappingRainWater.scala
object TrappingRainWater {
  def trap(height: Array[Int]): Int = {
    var leftMaxHeight = Array.empty[Int]
    var rightMaxHeight = Array.empty[Int]

    height.foldLeft(0)((item, h) => {
      val maxn = item max h
      leftMaxHeight = leftMaxHeight :+ maxn
      maxn
    })

    height.foldRight(0)((item, h) => {
      val maxn = item max h
      rightMaxHeight = maxn +: rightMaxHeight
      maxn
    })

    var water = 0
    for (i <- height.indices) {
      val minn = leftMaxHeight(i) min rightMaxHeight(i)

      if (minn > height(i)) {
        water += minn - height(i)
      }
    }

    water
  }

  def main(args: Array[String]): Unit = {
    val height = Array(0,1,0,2,1,0,1,3,2,1,2,1)
    println(trap(height))
  }
}
